#### UNDERSTANDING PYTHON PROGRAMMING $\;\;\;\;\;\;\;\;$   Created By: Siddhant Patel  $\;\;\;\;\;\;\;\;$   https://github.com/sid-patel 

# Chapter : Data Types in Python

#### Numeric data types: int, float, complex
#### String data types: str
#### Sequence types: list, tuple, range
#### Binary types: bytes, bytearray, memoryview
#### Mapping data type: dict
#### Boolean type: bool
#### Set data types: set, frozenset

#### Before jumping into Data Types in python,first lets understand some basic concepts in python :

 ### 1. Variable Assignment in Python

*  Variable in python is essentially a reference/name/label assigned or given to piece of data.

*  '=' is assignment operator. it means if age = 15 then 15 (value or data) is assigned (or stored) to variable called 'age'.

*  When we declare a variable in python, first python creates an object in memory to represent value (e.g. 15), then variable "age" becomes reference to that object. 

##### Do Variables Hold Memory?
##### Technically, variables themselves do not hold memory in Python. Instead:

* **Values or objects** are stored in memory.
* Variables are references (or pointers) to those memory locations.

##### If multiple variable reference the same object, then they share same memory until one of them is re-assigned. Shown in below example.

In [15]:
age = 15
print(age)
print(id(age)) # id is memory location of variable "age". so "age" is a variable(reference) that is pointing id(location) where object (or value) 15 is stored.

15
140721211413736


In [1]:
a = 10
b = 10
print(id(a))
print(id(b))

140721571206216
140721571206216


* Note : so as we can see here both the Variables a and b reference the same object 10 so they share the same memory location

##### Initializing multiple variables

In [6]:
Name,Id,Age = "Siddhant",1001,30
print("Name:",Name)
print("Id:",Id)
print("Age:",Age)

Name: Siddhant
Id: 1001
Age: 30


 ### 2. Memory Model in Python

  #### Python uses a reference counting mechanism for memory management. This is part of Python's Garbage Collection (GC) system. Here's how it works:
 * Every object in Python has a reference count (how many variables point to it).
 * When an object’s reference count drops to zero (i.e., no variable points to it), Python automatically frees that memory.

In [1]:
x = [1, 2, 3]
y = x       # `y` references the same list object
print(id(x))  # Same memory address for `x` and `y`
print(id(y))

1740289889024
1740289889024


In [3]:
y = None     # Now `y` no longer points to the list
print(x)     # `x` still references the list
print(y)
print(id(x)) 
print(id(y))

[1, 2, 3]
None
1740289889024
140721210026736


 ### 3. Dynamic Typing in Python
 #### Python is dynamically typed, meaning:

 * A variable can reference different types of objects at different times.
 * You don’t need to declare the type of a variable explicitly.

In [7]:
x = 25
print(x)
print(type(x))
print(id(x)) 

x = ('hello')
print(type(x))
print(id(x)) 
print(x)

25
<class 'int'>
140721211414056
<class 'str'>
1740268689456
hello


#### In Above code :
* first we assigned value 25 (integer) to variable x so for that data type of x is <class 'int'>
* then by reassigning x we changed value of x = 'hello' (string) so for that data type of x is <class 'str'>
#### Reassignment :
* When x is reassigned from 25 to 'hello', Python creates a new object for 'hello' in memory (at 1740268689456), and the reference to 25 is discarded (if nothing else references 25). id(x) changes because x points to a new object.

#### Dynamic Typed :
* Therefore variables are dynamically typed, they dont have data types but their data types depend upon data assigned / or stored in them. For example in c or c++ you assign data type while assigning variable ex. int x = 10 so here they are static. their data type wont change.
* Python is dynamically typed, meaning a variable can change its type during execution. Initially, x was an integer (int), but after reassignment, it became a string (str).


#### What Happens to Memory Location of 25?
* When x is reassigned to a new object ('hello'), the previous object (25) is unaffected in memory. However, its behavior depends on how many references are pointing to it.
##### 1. Reference count : 
##### In Python, objects are managed using a mechanism called reference counting:

* If no other variables or references point to 25, its reference count becomes 0.
* At this point, Python’s garbage collector marks the object (25) for cleanup, and its memory is reclaimed eventually.
* If another variable still points to 25, the object remains in memory.

##### Demonstration of Reference Count:
* You can manually inspect the reference count of an object using the sys.getrefcount() function:

In [11]:
import sys

print(sys.getrefcount(25)) # checking reference count for object or value 25.

x = 25
print(sys.getrefcount(25))  # Reference count for 25 increased by one as we assign one more reference x to the object 25. ( reference is usually higher due to internal references)

x = 'hello'                 # x no longer points to 25
print(sys.getrefcount(25))  # Reference count decreases by one as x is no longer points 25 because x points "hello" now. so reference counnt is decreased if no other references exist


1000000047
1000000048
1000000047


##### 2. Garbage Collection:

* When the reference count for an object drops to zero, **Python’s garbage collector** eventually cleans it up. This is not immediate but happens during the next garbage collection cycle.
* Python uses an internal garbage collection system to free up unused memory.


### 4. Immutable vs. Mutable Objects
* **Immutable objects (e.g., int, float, tuple, str)** cannot be changed after they are created. When you modify an immutable object, Python creates a new object in memory.
* **Mutable objects (e.g., list, dict, set)** can be changed without creating a new object.

#### Example : Immutable 

In [4]:
x = 10       # once integer 10 is assigned to x, it cant be changed
x += 1       # A new object is created for 11
print(x)

11


#### Example : Mutable

In [3]:
x = [1, 2, 3]
print(x)
print(id(x))
x.append(4)  # same list object can be modified and here append Modifies the list in-place
print(x)
print(id(x))

[1, 2, 3]
2259248504128
[1, 2, 3, 4]
2259248504128


### 5. Namespaces and Variable Scope
#### A namespace is a mapping of variable names to objects. Python has different types of namespaces:

* **Global Namespace:** Variables defined at the module level.
* **Local Namespace:** Variables defined inside functions.
* **Built-in Namespace:** Includes Python's built-in functions and objects.

In [35]:
x = 5  # Global variable

def Func():                  # defining function 
    x = 10  # Local variable
    print("Inside Func value of x is :", x)
    print(id(x))

s1 = Func()                       # calling function # we created s1 instance for it.
print("Outside Func value of x is :", x)
print(id(x))

Inside Func value of x is : 10
140728534602824
Outside Func value of x is : 5
140728534602664


* **Above Program Execution Flow**:
1. We assigned value 5 to variable x which is our Global Variable.
2. We have definded function called Func using inbuit keyword def that has variable x with value 10 and two print statements one by one.
3. As we are calling our Func() so Func() will be called and executed the output of execution is first two lines in our output.
4. Then both the print statements will be executed that are outside of func().

In [36]:
print(s1)

None


#### Key Observations:
**1. Global x :**
* The variable x = 5 is defined at the **global scope**, so it is accessible anywhere in the code outside the function Func().

**2. Local x :**

* The variable x = 10 is defined inside the Func() function, so it is a local variable and exists only during the function’s execution.
* **The local variable x inside the Func() function is not accessible outside the function.** This is because local variables have a **function scope** — they only exist and are accessible during the execution of the function in which they are defined. Once the function finishes execution, the local variable x is destroyed and cannot be accessed from outside.

**3. Function Return:**

* The function Func() does not return the local x, so **s1** is assigned the return value of **None** (default return value for a function with no return statement).


#### How to access Local x ?
##### Option 1: Return the Local Variable
* Modify the function to return the local x:

In [39]:
x = 5  # Global variable

def Func2():                  # defining function 
    x = 10  # Local variable
    print("Inside Func2 value of x is :", x)
    print(id(x))             # it will return MEMORY ID OF LOCAL x as an output
    return x                 # Return the Local variable

s2 = Func2()                  # s2 now stores the value of local x
print("Outside Func2 value of x is :", x)

Inside Func2 value of x is : 10
140728534602824
Outside Func2 value of x is : 5


In [40]:
print("Value of Local x stored in s2 is :",s2)

Value of Local x stored in s2 is : 10


##### Option 2: Store the Local Variable in a Class Attribute
* We can use a class to encapsulate the local variable as an attribute:

In [56]:

x = 5  # Global variable

class MyClass:
    def Func3(self):
        self.x = 10  # Store local x in an instance attribute
        print("Inside Func3 value of x is :", self.x)
        print("ID of self.x is:",id(self.x))

s3 = MyClass()
s3.Func3()

print("Access Local x via s3.x:", s3.x)
print("ID of Local x via s3.x:", id(s3.x))
print("Outside of Myclass value of global x is:",x)

Inside Func3 value of x is : 10
ID of self.x is: 140728534602824
Access Local x via s3.x: 10
ID of Local x via s3.x: 140728534602824
Outside of Myclass value of global x is: 5


In [57]:
s4 = MyClass()
s4.Func3()
print("ID of Local x via s4.x:", id(s4.x))

Inside Func3 value of x is : 10
ID of self.x is: 140728534602824
ID of Local x via s4.x: 140728534602824


##### Note:
* In Output **id(self.x)** ---> **140728534602824**  and **id(s3.x)** is also -----> **140728534602824**. It means they are same. **self.x = self.s3**
* It means **self = s3** for our s3 object (instance) when we assign or create **s3 = MyClass()**
  so now inside Func3 scenario is **s3.x = 10** so s3.x variable has value 10 so when we print(s3.x) output is 10
* Also when we create instance **s4 = Myclass()** and call function using **s4.Func3()** in that case our **self.x = s4.x = 10** because we can see
  the memory ID of both are same.
* Another interesting thing is **s3.x** and **s4.x** both are **different instances of MyClass has same Memory ID** ---> **140728534602824**. It is because the value stored is same which is 10. for example, when we assign a=10 , b = 10 then both a and b references holds same object 10 so both will point the same memory location where 10 is stored. its called **multiple variables assigned to same object**.(explained above in variable assignment) 

#### Lets see What happens when we change value of Local variable self.x for different instances by giving user input :

In [62]:
class MyClass2:
    def Func4(self):
        self.x = input("Provide value for Local x :")  # taking value of local x from user for every new instance of class
        self.y = input("Provide value for Local y :")
        print("Inside Func4 value of self.x is :", self.x)
        print("Inside Func4 value of self.y is :", self.y)
        print("ID of self.x is:",id(self.x))
        print("ID of self.y is:",id(self.y))
        
s5 = MyClass2()
s5.Func4()

print("Access Local self.x via s5.x:", s5.x)
print("Access Local self.y via s5.y:", s5.y)
print("ID of s5.x:", id(s5.x))
print("ID of s5.y:", id(s5.y))

Provide value for Local x : 15
Provide value for Local y : 20


Inside Func4 value of self.x is : 15
Inside Func4 value of self.y is : 20
ID of self.x is: 2259265753456
ID of self.y is: 2259265754544
Access Local self.x via s5.x: 15
Access Local self.y via s5.y: 20
ID of s5.x: 2259265753456
ID of s5.y: 2259265754544


In [64]:
s6 = MyClass2()
s6.Func4()
print("Access Local self.x via s6.x:", s6.x)
print("Access Local self.y via s6.y:", s6.y)
print("ID of s6.x:", id(s6.x))
print("ID of s6.y:", id(s6.y))

Provide value for Local x : 30
Provide value for Local y : 50


Inside Func4 value of self.x is : 30
Inside Func4 value of self.y is : 50
ID of self.x is: 2259265861104
ID of self.y is: 2259265859632
Access Local self.x via s6.x: 30
Access Local self.y via s6.y: 50
ID of s6.x: 2259265861104
ID of s6.y: 2259265859632


#### Note :
* For our s5 instance **self** will become **s5** so during **s5 = MyClass(), s5.func4()** we are **replacing self with s5** thats 
  why **id of self.x = id of s5.x** and **id of self.y = id of s5.y**
* Same way, for our s6 instance **self** will become **s6** so during **s6 = MyClass(), s6.func4()** we are **replacing self with s6** thats 
  why **id of self.x = id of s6.x** and **id of self.y = id of s6.y**
##### Here's importance of self :
* Using self we can create as many attributes as we want like self.x,self.y,self.g and so on. and with single instance for example s5 we will be able   to access all of them

#### Understanding "self" inbuilt keyword in Python with Real World Scenario :
* **Problem Statement :** Assume you're a teacher having 5 students in your class. 
* You are teaching them Math,Science and Physics and taking their monthly test for each subject of 100 Marks(each test).
* so after monthly exams, you have a sheet with 5 Columns. 
* In 1st column there are their names (Joy,David,Keval,Sid,Sara).
* In 2nd column "Obtained marks in Math out of 100"
* In 3nd column "Obtained marks in Science out of 100"
* In 4th column "Obtained marks in Physics out of 100"
* In 5th column is of "Percentage" that what you want to calculate using python.
* **Goal** is to calculate Percentage(%) for each student so whoever got highest percentage will be rewarded.
* So we will create a python program that will take an input Name,Math Score,Science Score,Physics Score and
  will give us percentage as an output for that student. so we will do the same for 5 times as we have five students.

In [67]:
class Student:
    def percentage(self):
        
        # Taking student inputs
        self.name = input("Enter the student's name: ")
        self.maths = float(input("Enter score for Maths (out of 100): "))
        self.science = float(input("Enter score for Science (out of 100): "))
        self.Physics = float(input("Enter marks for Physics (out of 100): "))
        
        # Calculating percentage
        total_marks = self.maths + self.science + self.Physics
        percentage = (total_marks / 300) * 100

        # Displaying the result
        print(f"\n{self.name}'s Percentage: {percentage:.2f}%")
# Creating an instance of the Student class
student1 = Student()
student2 = Student()
student3 = Student()
student4 = Student()
student5 = Student()

In [68]:
# Calling the percentage method
student1.percentage()

Enter the student's name:  Joy
Enter score for Maths (out of 100):  70
Enter score for Science (out of 100):  67.5
Enter marks for Physics (out of 100):  88



Joy's Percentage: 75.17%


In [69]:
student2.percentage()

Enter the student's name:  David
Enter score for Maths (out of 100):  56
Enter score for Science (out of 100):  80
Enter marks for Physics (out of 100):  98



David's Percentage: 78.00%


In [70]:
student3.percentage()

Enter the student's name:  Keval
Enter score for Maths (out of 100):  96
Enter score for Science (out of 100):  95
Enter marks for Physics (out of 100):  94



Keval's Percentage: 95.00%


In [71]:
student4.percentage()

Enter the student's name:  Sid
Enter score for Maths (out of 100):  88
Enter score for Science (out of 100):  86
Enter marks for Physics (out of 100):  85



Sid's Percentage: 86.33%


In [72]:
student5.percentage()

Enter the student's name:  sara
Enter score for Maths (out of 100):  78
Enter score for Science (out of 100):  75
Enter marks for Physics (out of 100):  76



sara's Percentage: 76.33%


#### Summary :
* So using **"Self"** keyword we can create as many attributes as we want in a function and assign or stores different values in them.
* Then we can create different objects or instances for that class and we will be able to access all those attributes and their values.

### Understanding Print Statement used in above program:
* #### print(f"\n{self.name}'s Percentage: {percentage:.2f}%")
* ##### Explanation:
1. f" at the Start:
* The f" indicates that this is an f-string, allowing expressions inside curly braces {} to be evaluated and inserted into the string dynamically.

2. \n:
* The \n adds a new line before the string output. This ensures the text starts on a new line.

3. {self.name}:
* Inside {}, the value of the self.name attribute (likely the student's name) is inserted dynamically.
* For example, if self.name = "Alice", this part of the string will be replaced with "Alice".

4. 's Percentage:
* A static part of the string, this will appear exactly as it is, providing context for the dynamic parts.

5. {percentage:.2f}:
* This embeds the value of the variable percentage into the string, formatting it to 2 decimal places.
* The .2f part is a format specifier:
    * f stands for "fixed-point notation."
    * .2 means the number will be rounded to 2 decimal places.
* For example:
    * If percentage = 85.666666, it will appear as 85.67.
    * If percentage = 92.0, it will appear as 92.00.
6. %:

* The % symbol at the end is a static part of the string, indicating the percentage unit.


### Understanding Design Philosophy of Python : "Everything in Python is an Object"
* ####  An integral part of Python's Object-Oriented Programming

In [73]:
class MyStudent:
    def percentage(self):
        # Displaying the type of self
        print(f"The data type of 'self' is: {type(self)}")

        # Taking user inputs
        self.name = input("Enter the student's name: ")
        self.maths = float(input("Enter marks for Maths (out of 100): "))
        self.science = float(input("Enter marks for Science (out of 100): "))
        self.english = float(input("Enter marks for English (out of 100): "))
        
        # Displaying the data type of self.name
        print(f"The data type of 'self.name' is: {type(self.name)}")
        print(f"The data type of 'self.maths' is: {type(self.maths)}")
        
        # Calculating percentage
        total_marks = self.maths + self.science + self.english
        percentage = (total_marks / 300) * 100
        
        # Printing the data type of percentage
        print(f"The data type of 'percentage' is: {type(percentage)}")
        print(f"The data type of 'total_marks' is: {type(total_marks)}")

        # Displaying the result
        print(f"\n{self.name}'s Percentage: {percentage:.2f}%")

# Creating an instance of the Student class
studentinput = MyStudent()

# Calling the percentage method
studentinput.percentage()

# Check the data type of class,an instance of class, the percentage function and 'percentage' when accessed from an instance

print(f"The data type of the 'def percentage' when accessed from the class is : {type(MyStudent.percentage)}")  # Accessing as a class attribute
print(f"The data type of 'studentinput.percentage()' method when accessed from instance is : {type(studentinput.percentage)}")  # Accessing as an instance attribute
print(f"The data type of an instace(studentinput) of a class MyStudent is : {type(studentinput)}")
print(f"The data type of 'class Mystudent' is : {type(MyStudent)}")

The data type of 'self' is: <class '__main__.MyStudent'>


Enter the student's name:  Siddhant Patel
Enter marks for Maths (out of 100):  92
Enter marks for Science (out of 100):  70
Enter marks for English (out of 100):  89


The data type of 'self.name' is: <class 'str'>
The data type of 'self.maths' is: <class 'float'>
The data type of 'percentage' is: <class 'float'>
The data type of 'total_marks' is: <class 'float'>

Siddhant Patel's Percentage: 83.67%
The data type of the 'def percentage' when accessed from the class is : <class 'function'>
The data type of 'studentinput.percentage()' method when accessed from instance is : <class 'method'>
The data type of an instace(studentinput) of a class MyStudent is : <class '__main__.MyStudent'>
The data type of 'class Mystudent' is : <class 'type'>
